### DISCUSSION TITLE
Open Recent VS Code Project

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#1238] [Share]
TIL VS Code has a sqlite database of recents, so I built this!

[Open open-recent-vs-code-project in Script Kit](https://scriptkit.com/api/new?name=open-recent-vs-code-project&url=https://gist.githubusercontent.com/johnlindquist/b2426f52a9b5f3ca4827fbdeda6b323c/raw/8d7cbd175540000bf6a8684814de265343ad2ae5/open-recent-vs-code-project.ts")

```js
// Name: Recent VS Code Project
// Cache: true
// Group: Favorite
// Keyword: vs

import "@johnlindquist/kit"
import { URL, fileURLToPath } from "url"

// /Users/johnlindquist/Library/Application Support/Code/User/globalStorage/state.vscdb
let filename = home("Library", "Application Support", "Code", "User", "globalStorage", "state.vscdb")
// windows path not tested, just guessing
if (isWin) filename = home("AppData", "Roaming", "Code", "User", "globalStorage", "state.vscdb")
let { default: sqlite3 } = await import("sqlite3")
let { open } = await import("sqlite")

let db = await open({
  filename,
  driver: sqlite3.Database,
})

let key = `history.recentlyOpenedPathsList`
let table = `ItemTable`

let result = await db.get(`SELECT * FROM ${table} WHERE key = '${key}'`)
let recentPaths = JSON.parse(result.value)
let recentFilePaths = []
for (let entry of recentPaths.entries) {
  if (entry?.folderUri && entry.folderUri.startsWith("file://")) {
    try {
      let path = fileURLToPath(new URL(entry.folderUri))
      recentFilePaths.push(path)
    } catch (error) {
      console.error(`Failed to parse ${entry.folderUri}. Error: ${error}`)
    }
  }
}

let recentPath = await arg("Open a recent path", recentFilePaths)
hide()
await exec(`code ${recentPath}`)
```

**Comment by shyagamzo** [comment]
Looks like this only works with certain paths...
```
[2023-05-04 11:09:09.571] [warn]  ☠️ ERROR PROMPT SHOULD SHOW ☠️
[2023-05-04 11:09:09.572] [warn]  TypeError [ERR_INVALID_URL_SCHEME]: The URL must be of scheme file
    at new NodeError (node:internal/errors:399:5)
    at fileURLToPath (node:internal/url:1492:11)
    at file:///C:/........................../.kenv/.scripts/open-recent-vs-code-project.mjs:18:89
    at Array.map (<anonymous>)
    at file:///C:/........................../.kenv/.scripts/open-recent-vs-code-project.mjs:18:76

```
I work with devcontainers and remote repos somtimes, so my `folderUri`s have different schemas, which breaks the URL conversion.
![image](https://user-images.githubusercontent.com/95415447/236145845-2e9da3dc-4584-4849-899c-9fcdd89793f5.png)

A quick fix is to filter for `startsWith('file://')`, but then I'm not getting my devcontainer and other projects.
Maybe other schemas will work too. Didn't try...

**Reply by mabry1985** [reply]
just installed 👍 love it!

**Reply by benschlegel** [reply]
I made a "v2" version of this script that fixes this issue (with most I tested like ssh/wsl connections (`vscode-remote://<path>` and virtual Github projects like `vscode-vfs://<path>`) and some more improvements like full windows support and more user friendly result display.

Link here: https://github.com/johnlindquist/kit/discussions/1418

(haven't tested it with dev containers as I don't have those set up, but those should probably also work)

**Comment by joshuaauger** [comment]
@johnlindquist could you update the site with the new code, when using the [Open open-recent-vs-code-project in Script Kit](https://scriptkit.com/api/new?name=open-recent-vs-code-project&url=https://gist.githubusercontent.com/johnlindquist/b2426f52a9b5f3ca4827fbdeda6b323c/raw/8d7cbd175540000bf6a8684814de265343ad2ae5/open-recent-vs-code-project.ts%22) link you show on [the site](https://www.scriptkit.com/johnlindquist/open-recent-vs-code-project), this code doesn't work, it references some old code.
